### ISSUE TITLE
Checkpoint and resume trainer-internal state (CMA-ES, DQN, REINFORCE)

### ISSUE THREAD

**Issue by espenhk** [espenhk/gamer-ai#57]
## Summary

Today the only thing persisted across runs is the **champion weight vector** in `policy_weights.yaml`. For stateful trainers this is not enough: interrupting a run and resuming silently restarts all *adaptation state* from scratch, even though `--re-initialize` was not passed. Add per-policy checkpointing for the trainer's internal state alongside the existing champion weights.

## Motivation

Stateful trainers that currently lose information on interrupt:

- **`CMAESPolicy`** loses the full search distribution: `mean`, `sigma`, covariance `C`, its factorization `(B, D, invsqrtC)`, evolution paths `ps` and `pc`, and the generation counter. On resume it restarts at the saved champion's weights but with `sigma = initial_sigma` and `C = I`, effectively undoing many generations of adaptation.
- **`NeuralDQNPolicy`** loses the replay buffer, the target network copy, and the SGD/Adam optimizer moments. Resuming behaves like a fresh DQN that happens to have a good initialization — target-network drift and replay warm-up have to happen again.
- **`REINFORCEPolicy`** loses the running-mean baseline and optimizer state.
- **`LSTMEvolutionPolicy`** is similar to CMA-ES — losing distribution state undoes adaptation even though the champion net is preserved.
- **`EpsilonGreedyPolicy` / `MCTSPolicy`** — Q-table persistence was already addressed in #12; these are called out for completeness so whatever mechanism we pick is consistent.

This matters operationally because long runs on Azure worker VMs (see `infrastructure/`) routinely get interrupted — cost-driven deallocation, RDP disconnects, network glitches during distributed grid search.

### File layout

Alongside `experiments/<track>/<name>/policy_weights.yaml`, write a sibling file:

```
experiments/<track>/<name>/trainer_state.npz   # numpy arrays + scalar metadata
```

`.npz` rather than YAML because the CMA-ES covariance is a float64 matrix that YAML serialisation would mangle.

For pure-Python state (optimizer hyperparams, counters) write a small companion YAML or store scalars in the same `.npz` via `np.savez(..., sigma=self._sigma, ...)`.

### API

Extend `BasePolicy` with two optional hooks (default no-op):

```python
class BasePolicy:
    def save_trainer_state(self, path: str) -> None: ...
    def load_trainer_state(self, path: str) -> None: ...
```

Each trainer overrides as needed:

- `CMAESPolicy.save_trainer_state` — `np.savez(path, mean=..., sigma=..., C=..., B=..., D=..., invsqrtC=..., ps=..., pc=..., gen=...)`.
- `NeuralDQNPolicy` — replay buffer (as a circular array) + target network weights + optimizer state.
- `REINFORCEPolicy` — baseline state + optimizer state.
- `LSTMEvolutionPolicy` — same shape as CMA-ES.

Call `save_trainer_state` after every improvement (where `save()` is already called), and `load_trainer_state` from `main.py` after loading the champion, unless `--re-initialize`.

### Cross-version safety

If the saved state is incompatible with the current code (e.g. `n` dimension changed because lidar ray count changed), fall back to "restart from champion weights, warn loudly". Do not silently discard. Concretely, validate shapes at load time and either raise or log-and-skip with a clear message.

### Grid search interaction

Distributed workers (`distributed/worker.py`) should copy back `trainer_state.npz` alongside the result payload so a coordinator-level resume works, not just single-machine resume. The simpler first pass can skip this: just make single-machine resume correct, and open a follow-up for distributed.

## Acceptance criteria

- [ ] Interrupting `python main.py <exp>` mid-run and re-running (without `--re-initialize`) continues with the saved CMA-ES `sigma` and covariance, not the defaults.
- [ ] Same for DQN replay buffer and target network.
- [ ] Same for REINFORCE baseline + optimizer state.
- [ ] `--re-initialize` deletes both `policy_weights.yaml` and `trainer_state.npz`.
- [ ] Shape-mismatch on load (e.g. `n_lidar_rays` changed) raises or logs a clear, actionable error — never silently continues with garbage state.
- [ ] Existing hill-climbing / genetic / tabular policies are unaffected (default no-op implementation).
- [ ] Unit test: save → reload → verify all arrays equal.

## Files to change

| File | Change |
|------|--------|
| `framework/policies.py` | Add `save_trainer_state` / `load_trainer_state` hooks to `BasePolicy` |
| `games/tmnf/policies.py` | Implement hooks for `CMAESPolicy`, `NeuralDQNPolicy`, `REINFORCEPolicy`, `LSTMEvolutionPolicy` |
| `framework/training.py` | Invoke `save_trainer_state` whenever champion weights are saved |
| `main.py` | Invoke `load_trainer_state` after champion load, unless `--re-initialize` |
| `tests/test_cmaes_policy.py` | Round-trip test for trainer state |

## Out of scope

- Persisting `GeneticPolicy` population state across runs. The genetic loop is already written so that resume-from-champion + a fresh population works fine; the value-per-effort ratio is lower than for CMA-ES.
- Cross-coordinator checkpointing in distributed mode (worth a separate follow-up after the single-machine story works).
